### ISSUE TITLE
[Question] How do I assign a global IP to a VM?

### ISSUE THREAD

**Issue by xeruf** [harvester/harvester#4129]
I researched and browsed issues for a while, but I cannot figure this out. I would like to give my VM an IP in the same range as the IPs of thq physical servers it is on.
The VM can ping the gateway and can be reached by its local IP, but regardless of which VM and which configuration tool, it does not get its global IP assigned and reachable from the outside.

I have no idea if I need to configure Cluster Networks, VM Networks, VLANs or whatever - I think the documentation on these topics is not very enlightening. What is the difference between masquerading or bridging the network? These config options are not explained anywhere, neither is there a link to another reference.

Basically, this is what I tried, mostly the same configuration as the host device, which  works smoothly:

    ip address add XX.XXX.XXX.93/27 broadcast + dev enp1s0
    ip route add XX.XXX.XXX.64/27 via XX.XXX.XXX.65 dev enp1s0

**Comment by xeruf** [comment]
By the way: As soon as I add that route, I cannot ping the gateway anymore. Odd.

**Comment by Martin-Weiss** [comment]
Not sure if I understand the question right - but giving a machine an IP is done via cloud-init or within the OS installed in the quest OS. You also might have to make sure to configure the network Interfaxe into the right vlan.

**Comment by xeruf** [comment]
Yes, I request the IP within the OS, everything seems fine, but the IP is not reachable from the outside.
I created a VLAN matching the outside Gateway and CIDR and assigned it as well, then the IP is reachable within the cluster (from the host) but not outside still.
How do I set the VLAN up correctly?

**Comment by Martin-Weiss** [comment]
@xeruf  - I am not sure if I get the network setup of your harvester correct. Do you have a management vlan and a separate VM vlan with using different network subnets and vlan IDs? Or you you "just" have the management vlan?

**Comment by xeruf** [comment]
I tried both.
With just the management vlan, the IP was not reachable from even the host.
With a separate VM vlan configured with the Gateway and CIDR of the outside router, it was reachable from the host but not from outside.

**Comment by xeruf** [comment]
Is it correct that the VM IP in the Web UI (10.52.0.117) is pingable yet differs from the IP as shown by `ip a` within the VM, in this case 10.0.2.2 which is not reachable even from the host?

**Comment by xeruf** [comment]
Working now, my steps:
- create a VM Network set to "Cluster Network: mgmt", VLAN tag does not matter
- assign that device to the VM
- assign the IP to that device (which is the second one!) with a default route via the gateway:
```sh
ip address add XX.XXX.XXX.93/27 broadcast + dev enp2s0
ip route add default via XX.XXX.XXX.65 dev enp2s0
```

**Comment by xeruf** [comment]
Would be prudent to add it to the docs, for people coming to this platform without a deep understanding of the mechanisms like me :)

**Comment by nastynaz** [comment]
I'd also greatly appreciate it if the docs on networking were updated. More clarification on cluster networks, VIPs and how the VLANs work with harvester would be really helpful.

**Comment by cruno91** [comment]
> Working now, my steps:
> 
> * create a VM Network set to "Cluster Network: mgmt", VLAN tag does not matter
> * assign that device to the VM
> * assign the IP to that device (which is the second one!) with a default route via the gateway:
> 
> ```shell
> ip address add XX.XXX.XXX.93/27 broadcast + dev enp2s0
> ip route add default via XX.XXX.XXX.65 dev enp2s0
> ```

Thanks for getting this far.  I have been searching for a few days now on how to achieve this so I can access my VMs from other physical machines.

I'm stuck on the last command there though.  

I'm trying to assign an IP (10.0.3.123) on my physical network (10.0.3.X/24).  

I have created the network in Harvester:
<img width="718" alt="Screenshot 2023-08-20 at 11 34 25 AM" src="https://github.com/harvester/harvester/assets/1760366/df8d129a-42bc-4f02-abd4-a924eb001c55">


I attached the network to the VM:
<img width="1409" alt="Screenshot 2023-08-20 at 11 35 18 AM"
 src="https://github.com/harvester/harvester/assets/1760366/a4f0e8f4-ba1d-4a47-843f-4986380a0548">

I ran the first command `ip address add 10.0.3.123/24 broadcast + dev enp2s0` successfull, but the second command `ip route add default via 10.0.3.1 dev enp2s0` fails with the error `Nexthop has invalid gateway`:
<img width="815" alt="Screenshot 2023-08-20 at 11 38 25 AM" src="https://github.com/harvester/harvester/assets/1760366/6dbe99bc-b392-4b79-b25e-ad699df216d2">

So I'm not really sure what I'm doing wrong.

This is the config for my physical network if it helps:
<img width="763" alt="Screenshot 2023-08-20 at 11 49 56 AM" src="https://github.com/harvester/harvester/assets/1760366/dcf20efc-801a-46ba-9283-80a79f47e568">

**SOLVED**

The VLAN in Harvester had to have the same tag ID as the physical network VLAN the Harvester box is on (3 in my case).
Then I just had to create a netplan config (in the Ubuntu VM) which grabs a fixed IP from the reserved pool on my network's VLAN.

Alternatively I also was able to use an untagged VLAN in Harvester to just grab from the DHCP pool on my physical network's VLAN.
